lib/ansible/module_utils/fortios.py

```diff
@@ -198,3 +208,819 @@ def get_error_infos(self, cli_errors):
+        if isinstance(path, list):
+            path = '/'.join(path) + '/'
+        return requests.post(self._ip + '/api/' + api + '/' + path, headers=self.header, cookies=self.cookies,
+                             verify=self._verify, proxies=self.proxies, params=params, json={'json': data})
+
+    @connection_handler
+    def _delete(self, path, api='v2', params=None, data=None):
+        if isinstance(path, list):
+            path = '/'.join(path) + '/'
+        return requests.delete(self._ip + '/api/' + api + '/' + path, headers=self.header,
+                               cookies=self.cookies, verify=self._verify, proxies=self.proxies, params=params,
+                               json={'json': data})
+
+    @return_handler
+    def _show(self, path, api='v2', params=None):
+        return self._get(path, api=api, params=self._build_params(params)).json()
+
+    @return_handler
+    def _edit(self, path, api='v2', params=None, data=None):
+        return self._put(path, api=api, params=self._build_params(params), data=data).json()
+
+    @return_handler
+    def _create(self, path, api='v2', params=None, data=None):
+        return self._post(path, api=api, params=self._build_params(params), data=data).json()
+
+    @return_handler
+    def _remove(self, path, api='v2', params=None, data=None):
+        return self._delete(path, api=api, params=self._build_params(params), data=data).json()
+
+    def _build_params(self, new_params):
+        params = {}
+        if self._vdom:
+            params['vdom'] = self._vdom
+        if new_params:
+            params.update(new_params)
+
+        if params:
+            return params
+        else:
+            return None
+
+    def _get_local_spec(self):
+        try:
+            with open(self._arg_spec_filename, "r+") as f:
+                local_spec = json.load(f)
+        except IOError:
+            local_spec = None
+        return local_spec
+
+    def _write_local_spec(self, spec):
+        with open(self._arg_spec_filename, "w+") as f:
+            json.dump(spec, f, indent=4, sort_keys=True)
+
+    def _get_permanent_object_identifiers(self):
+        perm_objs_from_module = self._api_info.get("permanent_objects", [])
+        perm_objs_from_config = self._params.get("permanent_objects", [])
+        if perm_objs_from_module is None:
+            perm_objs_from_module = []
+        if perm_objs_from_config is None:
+            perm_objs_from_config = []
+        return perm_objs_from_config + perm_objs_from_module
+
+    def _get_ignore_object_identifiers(self):
+        perm_objs_from_module = self._api_info.get("ignore_objects", [])
+        perm_objs_from_config = self._params.get("ignore_objects", [])
+        if perm_objs_from_module is None:
+            perm_objs_from_module = []
+        if perm_objs_from_config is None:
+            perm_objs_from_config = []
+        return perm_objs_from_config + perm_objs_from_module
+
+    def _get_argument_spec(self):
```

Comment:
This function is making an API call to get schema for the API endpoint and makes an argument spec for the given endpoint out of it. For example:
    "cmdb/firewall/addrgrp": {
        "arg_spec": {
            "allow-routing": {
                "options": [
                    "enable", 
                    "disable"
                ], 
                "type": "str"
            }, 
            "color": {
                "type": "int"
            }, 
            "comment": {
                "type": "str"
            }, 
            "member": {
                "options": {
                    "name": {
                        "type": "str"
                    }
                }, 
                "type": "list"
            }, 
            "name": {
                "type": "str"
            }, 
            "tags": {
                "options": {
                    "name": {
                        "type": "str"
                    }
                }, 
                "type": "list"
            }, 
            "uuid": {
                "type": "str"
            }, 
            "visibility": {
                "options": [
                    "enable", 
                    "disable"
                ], 
                "type": "str"
            }
        }, 
        "default_object": {
            "allow-routing": "disable", 
            "color": 0, 
            "comment": "", 
            "css-class": "ftnt-address-group ftnt-color-0", 
            "member": [], 
            "name": "", 
            "q_origin_key": "", 
            "tags": [], 
            "uuid": "00000000-0000-0000-0000-000000000000", 
            "visibility": "enable"
        }
    },

Reply:
I think that makes me even more concerned.  You are saying that the argument spec that ansible will use becomes dynamic based on responses of APIs?

I'm pretty sure that's not going to be allowed.

Reply:
We have historically not allowed generic API modules.  We generally require a module to be a specific set of features that perform an action.  This would be in opposition to that stand point.

Reply:
That is correct, argument spec is dynamic. If it is not dynamic, then any change to the API would render this module useless. Also, it would require me to write modules for every single API endpoint. There are some specific modules - https://github.com/willwagner602/ansible/tree/will-devel/lib/ansible/modules/network/fortios but there are a lot more of endpoints. So it made sense to let this core module build dynamic argument spec based on the data received from the API.

Reply:
And just to clarify - currently the argument spec per API endpoint is being cached locally. So if API changes, the file has to be deleted and then it can be recreated again.
Here's an example of this file after running a few playbooks: https://pastebin.com/CcPuMQCu

Reply:
Due to this, at the moment I am denying acceptance of this PR.  I'll ask to get additional feedback from some other core developers, but we have explicitly denied modules that attempted to address arbitrary API functionality like this in the past.

The recommended way would be to write individual modules for individual parts of functionality.

Reply:
OK, I feel like it makes sense for the modules that are using screen scraping via SSH, but I see a lot of value of treating API calls as generic as possible. I'll hold off on making any changes to the module until the core members provide feedback. I'll be happy to attend the meeting to explain my point of view to the core developers.